File: poc/c++_extensions/song/src/song.cpp
#include <iostream>
#include <string>

#include "song.h"


Song::Song(const std::string& name, double length) :
    m_name(name), m_length(length)
{}


void Song::SetName(const std::string& name)
{
    this->m_name = name;
}


void Song::SetLength(double length)
{
    this->m_length = length;
}


std::string Song::GetName() const
{
    return this->m_name;
}


double Song::GetLength() const
{
    return this->m_length;
}


void Song::PrintMessage() const
{
    std::cout << "Hello from " << this->m_name << std::endl;
}

File: poc/c++_extensions/song/wrapper/song.py
try:
    import cSong
except:
    strErr = "\n\n`cSong` module not found, "
    raise RuntimeError(strErr)


def PrintFunctionalMessage(message):
    cSong.PrintFunctionalMessage(message)


class song():
    def __init__(self, name, length):
        self.songCapsule = cSong.construct(name, length)

    def __delete__(self):
        cSong.delete_object(self.songCapsule)

    def SetName(self, name):
        cSong.SetName(self.songCapsule, name)

    def SetLength(self, length):
        cSong.SetLength(self.songCapsule, length)

    def GetName(self):
        return cSong.GetName(self.songCapsule)

    def GetLength(self):
        return cSong.GetLength(self.songCapsule)

    def PrintMessage(self):
        cSong.PrintMessage(self.songCapsule)
